File: apps/moot-bridge/Sources/moot-bridge/JSONValue.swift
import Foundation

// JSONValue.swift — the loosely-typed JSON value the bridge uses to parse and
// rebuild JSON-RPC messages whose shapes are not known at compile time.
//
// Carried over verbatim (our code) from the proven benchmarker skeleton
// (tools/mcp-benchmarker/.../MCPClient.swift). The bridge is engine-agnostic: it
// does not know the two backends' tool names or result shapes at compile time,
// so it parses just enough of each message to classify + translate it and
// re-encodes the rest. A full typed model would couple the bridge to one server's
// schema; this stays decoupled.

/// An error raised while talking to an MCP backend.
struct MCPError: Error, Sendable, CustomStringConvertible {
    let description: String
}

/// A loosely-typed JSON value, used both to build tool-call arguments and to
/// parse tool results from servers whose result shapes are not known at compile
/// time (the bridge is engine-agnostic).
enum JSONValue: Codable, Sendable, Equatable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
        } else if let b = try? c.decode(Bool.self) {
            self = .bool(b)
        } else if let n = try? c.decode(Double.self) {
            self = .number(n)
        } else if let s = try? c.decode(String.self) {
            self = .string(s)
        } else if let a = try? c.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? c.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(
                in: c, debugDescription: "unsupported JSON value")
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .null: try c.encodeNil()
        case .bool(let b): try c.encode(b)
        case .number(let n): try c.encode(n)
        case .string(let s): try c.encode(s)
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        }
    }

    /// The value at an object key, or nil if not an object / key absent.
    subscript(key: String) -> JSONValue? {
        if case .object(let o) = self { return o[key] }
        return nil
    }

    /// The string payload, if this value is a string.
    var stringValue: String? {
        if case .string(let s) = self { return s }
        return nil
    }

    /// The object payload, if this value is an object.
    var objectValue: [String: JSONValue]? {
        if case .object(let o) = self { return o }
        return nil
    }

    /// The array payload, if this value is an array.
    var arrayValue: [JSONValue]? {
        if case .array(let a) = self { return a }
        return nil
    }
}

File: benchmarks/rust/src/json_value.rs
//! json_value.rs — a loosely-typed JSON value.
//!
//! Ports Swift `JSONValue` (in MCPClient.swift). Used both to build
//! tool-call arguments and to parse tool results from servers whose result
//! shapes are not known at compile time (the benchmarker is engine-agnostic).
//!
//! Encoding/decoding rides on `serde_json::Value` so the wire bytes match the
//! Swift `JSONEncoder`/`JSONDecoder` output for the same logical value: an
//! object is a JSON object, a number is a JSON number, etc. The one wrinkle
//! the Swift leg has — encoding request ids as whole numbers (`100`, not
//! `100.0`) — is preserved here because integral `f64`s serialize through
//! serde_json as integers.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

/// A loosely-typed JSON value. Mirrors Swift `JSONValue`.
///
/// Object members are stored in a `BTreeMap` so iteration order is
/// deterministic (lexicographic by key) — the Swift leg sorts object keys at
/// the points where order matters (`objectArray`), and a stable order here
/// makes encode output reproducible across runs and matches that discipline.
#[derive(Debug, Clone, PartialEq)]
pub enum JsonValue {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<JsonValue>),
    Object(BTreeMap<String, JsonValue>),
}

impl JsonValue {
    /// The value at an object key, or `None` if not an object / key absent.
    /// Mirrors Swift `JSONValue.subscript(key:)`.
    pub fn get(&self, key: &str) -> Option<&JsonValue> {
        match self {
            JsonValue::Object(map) => map.get(key),
            _ => None,
        }
    }

    /// The string payload, if this value is a string. Mirrors Swift
    /// `JSONValue.stringValue`.
    pub fn string_value(&self) -> Option<&str> {
        match self {
            JsonValue::String(s) => Some(s.as_str()),
            _ => None,
        }
    }

    /// Convenience constructor for an object from key/value pairs.
    pub fn object<I>(pairs: I) -> JsonValue
    where
        I: IntoIterator<Item = (String, JsonValue)>,
    {
        JsonValue::Object(pairs.into_iter().collect())
    }

    /// Parses raw JSON bytes into a `JsonValue`. Mirrors decoding a
    /// `JSONValue` via `JSONDecoder` on the Swift leg.
    pub fn from_slice(data: &[u8]) -> Result<JsonValue, serde_json::Error> {
        let v: Value = serde_json::from_slice(data)?;
        Ok(JsonValue::from(v))
    }

    /// Serializes this value to JSON bytes. Mirrors encoding a `JSONValue`
    /// via `JSONEncoder` on the Swift leg.
    pub fn to_vec(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(&Value::from(self.clone()))
    }
}

// ── Bridge to/from serde_json::Value ─────────────────────────────────────────
// We round-trip through serde_json::Value so the wire format is byte-compatible
// with the Swift Foundation encoder for the same logical value.

impl From<Value> for JsonValue {
    fn from(v: Value) -> Self {
        match v {
            Value::Null => JsonValue::Null,
            Value::Bool(b) => JsonValue::Bool(b),
            Value::Number(n) => JsonValue::Number(n.as_f64().unwrap_or(0.0)),
            Value::String(s) => JsonValue::String(s),
            Value::Array(a) => JsonValue::Array(a.into_iter().map(JsonValue::from).collect()),
            Value::Object(o) => {
                JsonValue::Object(o.into_iter().map(|(k, v)| (k, JsonValue::from(v))).collect())
            }
        }
    }
}

impl From<JsonValue> for Value {
    fn from(v: JsonValue) -> Self {
        match v {
            JsonValue::Null => Value::Null,
            JsonValue::Bool(b) => Value::Bool(b),
            JsonValue::Number(n) => {
                // Encode an integral f64 as a JSON integer (100, not 100.0) so
                // the wire matches Swift's encoder for whole-number request ids
                // and pagination args — the servers accept integers, not floats.
                if n.fract() == 0.0 && n.is_finite() && n.abs() < 9.007_199_254_740_992e15 {
                    Value::Number((n as i64).into())
                } else {
                    serde_json::Number::from_f64(n)
                        .map(Value::Number)
                        .unwrap_or(Value::Null)
                }
            }
            JsonValue::String(s) => Value::String(s),
            JsonValue::Array(a) => Value::Array(a.into_iter().map(Value::from).collect()),
            JsonValue::Object(o) => {
                Value::Object(o.into_iter().map(|(k, v)| (k, Value::from(v))).collect())
            }
        }
    }
}

// Serde impls delegate to the serde_json::Value bridge so callers can embed a
// JsonValue inside other serde structures if needed.
impl Serialize for JsonValue {
    fn serialize<S: serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
        Value::from(self.clone()).serialize(s)
    }
}

impl<'de> Deserialize<'de> for JsonValue {
    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
        Value::deserialize(d).map(JsonValue::from)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_on_object_returns_member() {
        let v = JsonValue::object([("k".to_string(), JsonValue::String("v".to_string()))]);
        assert_eq!(v.get("k").and_then(|x| x.string_value()), Some("v"));
        assert!(v.get("missing").is_none());
    }

    #[test]
    fn get_on_non_object_is_none() {
        assert!(JsonValue::Bool(true).get("k").is_none());
    }

    #[test]
    fn integral_number_encodes_without_decimal() {
        let v = JsonValue::Number(100.0);
        let bytes = v.to_vec().unwrap();
        assert_eq!(String::from_utf8(bytes).unwrap(), "100");
    }

    #[test]
    fn round_trip_object() {
        let bytes = br#"{"a":1,"b":"x","c":[true,null]}"#;
        let v = JsonValue::from_slice(bytes).unwrap();
        assert_eq!(v.get("b").and_then(|x| x.string_value()), Some("x"));
        match v.get("c") {
            Some(JsonValue::Array(a)) => assert_eq!(a.len(), 2),
            other => panic!("expected array, got {other:?}"),
        }
    }
}
